.github: Enable SSH to linux runners

seemethere opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62280

Enables SSH to linux GHA runners for FB employees while on the FB VPN

SSH keys will be added to runners when the label "with-ssh" is applied to
your pull request.

Depends on https://github.com/fairinternal/pytorch-gha-infra/pull/8

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Closes https://github.com/fairinternal/pytorch-gha-infra/issues/5

Differential Revision: [D29941681](https://our.internmc.facebook.com/intern/diff/D29941681)

seemethere committed:
.github: Enable SSH to linux runners

Enables SSH to linux GHA runners for FB employees while on the FB VPN

SSH keys will be added to runners when the label "with-ssh" is applied to
your pull request.

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

[ghstack-poisoned]

seemethere commented:
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29941681).

seemethere committed:
Update on ".github: Enable SSH to linux runners"


Enables SSH to linux GHA runners for FB employees while on the FB VPN

SSH keys will be added to runners when the label "with-ssh" is applied to
your pull request.

Depnds on https://github.com/fairinternal/pytorch-gha-infra/pull/8

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29941681](https://our.internmc.facebook.com/intern/diff/D29941681)

[ghstack-poisoned]

seemethere committed:
Update on ".github: Enable SSH to linux runners"


Enables SSH to linux GHA runners for FB employees while on the FB VPN

SSH keys will be added to runners when the label "with-ssh" is applied to
your pull request.

Depnds on https://github.com/fairinternal/pytorch-gha-infra/pull/8

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29941681](https://our.internmc.facebook.com/intern/diff/D29941681)

[ghstack-poisoned]

seemethere committed:
Update on ".github: Enable SSH to linux runners"


Enables SSH to linux GHA runners for FB employees while on the FB VPN

SSH keys will be added to runners when the label "with-ssh" is applied to
your pull request.

Depnds on https://github.com/fairinternal/pytorch-gha-infra/pull/8

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29941681](https://our.internmc.facebook.com/intern/diff/D29941681)

[ghstack-poisoned]

seemethere committed:
Update on ".github: Enable SSH to linux runners"


Enables SSH to linux GHA runners for FB employees while on the FB VPN

SSH keys will be added to runners when the label "with-ssh" is applied to
your pull request.

Depnds on https://github.com/fairinternal/pytorch-gha-infra/pull/8

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29941681](https://our.internmc.facebook.com/intern/diff/D29941681)

[ghstack-poisoned]

janeyx99 commented on .github/templates/linux_ci_workflow.yml.j2:
```diff
@@ -231,6 +237,19 @@ jobs:
+        # Always hold for active ssh sessions
+        if: always()
+        run: |
+          echo "Holding runner for 2 hours until all ssh sessions have logged out"
```
This means that even if the ssh session is on, the runner will still shut down after 2 hours regardless, right?

seemethere replied:
Yes, this is consistent with CircleCI's current re-run with ssh functionality, see: https://circleci.com/docs/2.0/ssh-access-jobs/

samestep commented on .github/templates/linux_ci_workflow.yml.j2:
```diff
@@ -564,6 +606,19 @@ jobs:
           name: pytorch_github_io
           if-no-files-found: error
           path: pytorch_github_io.zip
+      - name: Hold runner for 2 hours or until ssh sessions have drained
+        # Always hold for active ssh sessions
+        if: always()
+        run: |
+          echo "Holding runner for 2 hours until all ssh sessions have logged out"
+          for _ in $(seq 1440); do
+            # Break if no ssh session exists anymore
+            if [[ "$(who)" = "" ]]; then
+              break
+            fi
+            echo "."
+            sleep 5
+          done
```
looks like this same `run` script is used in a few different places; would it make sense to put it into `.github/scripts` and just call it from here?

samestep reviewed (approved):
lgtm! I guess most of the logic is in https://github.com/seemethere/add-github-ssh-key?

seemethere commented:
> lgtm! I guess most of the logic is in [seemethere/add-github-ssh-key](https://github.com/seemethere/add-github-ssh-key)?

Yup most of the logic for this particular feature is in the terraform for the security groups as well as the the github actions to add the ssh keys.

seemethere committed:
Update on ".github: Enable SSH to linux runners"


Enables SSH to linux GHA runners for FB employees while on the FB VPN

SSH keys will be added to runners when the label "with-ssh" is applied to
your pull request.

Depends on https://github.com/fairinternal/pytorch-gha-infra/pull/8

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Closes https://github.com/fairinternal/pytorch-gha-infra/issues/5

Differential Revision: [D29941681](https://our.internmc.facebook.com/intern/diff/D29941681)

[ghstack-poisoned]

seemethere commented:
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29941681).

seemethere committed:
Update on ".github: Enable SSH to linux runners"


Enables SSH to linux GHA runners for FB employees while on the FB VPN

SSH keys will be added to runners when the label "with-ssh" is applied to
your pull request.

Depends on https://github.com/fairinternal/pytorch-gha-infra/pull/8

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Closes https://github.com/fairinternal/pytorch-gha-infra/issues/5

Differential Revision: [D29941681](https://our.internmc.facebook.com/intern/diff/D29941681)

[ghstack-poisoned]

seemethere commented:
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29941681).

JpD9119 commented:
[เปิดใช้งาน SSH เพื่อรันลินุกซ์ GHA](https://github.com/pytorch/pytorch.wiki.git) 

> ``
